### ISSUE TITLE
contigency_plan

### ISSUE THREAD

**Issue by mr-d-self-driving** [OpenDriveLab/BeTop#11]
Great job，No such py file?

**Comment by georgeliu233** [comment]
Hi, its in this file:
https://github.com/OpenDriveLab/BeTop/blob/nuplan-fixed/planning/src/models/betop/modules/contigency_utils.py

**Comment by mr-d-self-driving** [comment]
Great job，thanks a lot ！

**Comment by mr-d-self-driving** [comment]
> Hi, its in this file: https://github.com/OpenDriveLab/BeTop/blob/nuplan-fixed/planning/src/models/betop/modules/contigency_utils.py

Sorry，no ContigencyPlanner  code in  BeTop/planning/src/planners/vanilla_example.py file。

**Comment by georgeliu233** [comment]
oh, that one is a deprecated file. Previously I tried to use Theasus to perform the contigency planning, but it failed. So I simply make it as a loss function during training. I leave the code here for Theasus implementation, and you may have a try after some proper adjustment

```
import torch
import theseus as th

class ContigencyPlanner:
    def __init__(self, 
        device, 
        feature_len=9, 
        horizon=80, 
        branch_horizon=50,
        modal=6,
        num_neighbors=32, 
        test=False):   
        # cost_function_weights = [th.ScaleCostWeight(th.Variable(torch.rand(1), name=f'cost_function_weight_{i+1}'))
        #                          for i in range(feature_len)]

        index_weight = [th.Variable(torch.empty((1, 1)), name=f'index_{i}') for i in range(modal)]
            
        joint_plan = th.Vector(dof=branch_horizon*2, name='joint_plan')
        sep_plan = th.Vector(dof=modal*(horizon - branch_horizon)*2, name='sep_plan')
        actors = th.Variable(torch.empty(1, num_neighbors, modal, horizon, 3), name="actors")
        actors_hw = th.Variable(torch.empty(1, num_neighbors, 2), name="actors_hw")
        mode_score = th.Variable(torch.empty(1, modal), name="mode_score")
        actor_mask = th.Variable(torch.empty(1, num_neighbors), name="actor_mask")

        original_plan = th.Variable(torch.empty(1, horizon, 2), name="original_plan")

        self.branch_horizon = branch_horizon
        self.modal = modal
        self.horizon = horizon
        self.num_neighbors = num_neighbors

        # ref_line = th.Variable(torch.empty(1, 1000, 5), name="ref_line")
        # current_state = th.Variable(torch.empty(1, num_neighbors+1, 9), name="current_state")

        objective = th.Objective()
        objective = self.cost_function(
            objective, 
            joint_plan, sep_plan, mode_score,
            actors, actors_hw, original_plan,
            actor_mask, index_weight)
    
        if test:
            self.optimizer = th.LevenbergMarquardt(objective, th.CholeskyDenseSolver, max_iterations=30,
                                            step_size=0.3, vectorize=False, abs_err_tolerance=1e-2)
        else:
            self.optimizer = th.LevenbergMarquardt(objective, th.CholeskyDenseSolver, max_iterations=10, 
                                            step_size=0.5, vectorize=False)

        self.layer = th.TheseusLayer(self.optimizer, vectorize=False)
        self.device = device
        self.test = test
        self.layer.to(device)
    
    def inverse_dynamics(self, )
    
    def potential_fields(self, src_plan, actor, actor_hw):
        '''
        src_plan: [b, t, 2]
        actor: [b, A, t, 3] (x, y, theta)
        actor_hw: [b, A, 2] (height, width)
        '''
        theta = actor[..., -1]
        cos, sin = torch.cos(theta), torch.sin(theta)
        diff = actor[..., :2] - src_plan[:, None, :, :2]
        dx, dy = diff[..., 0], diff[..., 1]
        # new_dx = cos * dx + sin * dy 
        # new_dy = -sin * dx + cos * dy

        h, w = 1 / actor_hw[..., None, 0], 1 / actor_hw[..., None, 1]

        field_cost = 1/ (1 +torch.square(dx) + torch.square(dy))
        
        return field_cost

    
    def joint_safety_cost(self, optim_vars, aux_vars):
        src_joint_plan = optim_vars[0].tensor.view(-1, self.branch_horizon, 2)
        actors = aux_vars[0].tensor
        actor_hw = aux_vars[1].tensor
        actor_mask = aux_vars[2].tensor

        b, a, m, t, d = actors.shape
        actors = actors.view(b, a*m, t ,d)
        actors = actors[..., :self.branch_horizon, :]
        actor_hw = actor_hw.unsqueeze(2).repeat(1, 1, m, 1)
        actor_hw = actor_hw.view(b, a*m, 2)

        # selected_horizon = [1, 2, 4, 9, 14, 19, 24, 29, 39, 49]
        src_joint_plan = src_joint_plan[:, :, :]
       
        actors = actors[:, :, :, :]

        #[b, a*m, sel_horizon]
        field_cost = self.potential_fields(src_joint_plan, actors, actor_hw)
        actor_mask = actor_mask[:, :, None].repeat(1, 1, m).view(b ,a*m)
        field_cost = field_cost - (1 - actor_mask)[:, :, None] * 10e5

        #[b, sel_horizon]: the one with minimum scaled dist to ego with all modes and actors
        worst_case_cost = torch.max(field_cost, dim=1)[0]
        return 10 * worst_case_cost
    
    def marginal_safety_cost(self, optim_vars, aux_vars):
        src_joint_plan = optim_vars[0].tensor.view(-1, self.modal, self.horizon - self.branch_horizon, 2)
        actors = aux_vars[0].tensor
        actor_hw = aux_vars[1].tensor
        mode_weights = aux_vars[2].tensor
        mode_index = aux_vars[3].tensor[0, 0].long()
        actor_mask = aux_vars[4].tensor

        src_marginal_plan = src_joint_plan[:, mode_index]
        b, a, m, t, d = actors.shape
        actors = actors[..., self.branch_horizon:, :]
        actors = actors[:, :, mode_index, :, :]

        # selected_horizon = [9, 19, 29]
        src_marginal_plan = src_marginal_plan[:, :, :]
        actors = actors[..., :, :]

        #[b, a, sel_horizon]
        field_cost = self.potential_fields(src_marginal_plan, actors, actor_hw)
        field_cost = field_cost - (1 - actor_mask)[:, :, None] * 10e5

        #[b, sel_horizon]: the one with minimum scaled dist to ego with all actors
        worst_case_cost = torch.max(field_cost, dim=1)[0]
        cost =  mode_weights[:, mode_index, None] * worst_case_cost
        return 10 * cost
    
    def joint_il_cost(self, optim_vars, aux_vars):
        src_joint_plan = optim_vars[0].tensor.view(-1, self.branch_horizon, 2)
        original_plan = aux_vars[0].tensor
        joint_cost = (src_joint_plan - original_plan[:, :self.branch_horizon])[:, :]
        b = joint_cost.shape[0]

        return joint_cost.reshape(b, -1)
    
    def sep_il_cost(self, optim_vars, aux_vars):
        src_joint_plan = optim_vars[0].tensor.view(-1, self.modal, self.horizon - self.branch_horizon, 2)
        original_plan = aux_vars[0].tensor
        mode_weights = aux_vars[1].tensor
        mode_index = aux_vars[2].tensor[0, 0].long()

        src_marginal_plan = src_joint_plan[:, mode_index]
        err = (src_marginal_plan - original_plan[:, self.branch_horizon:])#[:, ::2]

        xy_err = mode_weights[:, mode_index, None, None] * err
        batch = xy_err.shape[0]

        return xy_err.reshape(batch, -1)


    def cost_function(
        self, 
        objective, 
        joint_plan, sep_plan, mode_score,
        actors, actors_hw, original_plan,
        actor_mask, index_weight,
        vectorize=True
        ):

        joint_safe_cost = th.AutoDiffCostFunction(
                    [joint_plan], self.joint_safety_cost, 50, #cost_weight=5.,
                    aux_vars=[actors, actors_hw, actor_mask], 
                    autograd_vectorize=vectorize, name="joint_safe_cost")
        objective.add(joint_safe_cost)

        for i in range(self.modal):
            marginal_safety_cost = th.AutoDiffCostFunction(
                        [sep_plan], self.marginal_safety_cost, 30, #cost_weight=5.,
                        aux_vars=[actors, actors_hw, mode_score, index_weight[i], actor_mask], 
                        autograd_vectorize=vectorize, name=f"sep_safe_cost_{i}")
            objective.add(marginal_safety_cost)
        
        return objective
    
    def plan(self, data, out):
        
        actor_shape = data["agent"]["shape"][:, 1:, 20, :2]
        actor_curr_pos = data["agent"]["position"][:, 1:, 20, :2]

        actor_preds = out['prediction']
        actor_score = out['pred_probability'].sum(1).softmax(-1)

        actor_pred_pos = actor_preds[..., :2] + actor_curr_pos[:, :, None, None]
        actor_angle = torch.atan2(actor_preds[..., 3], actor_preds[..., 2])

        actor_pred = torch.cat([actor_pred_pos, actor_angle.unsqueeze(-1)], dim=-1)

        actor_mask = torch.any(data["agent"]["valid_mask"][:, 1:, :20], dim=-1)
        actor_mask = actor_mask.float()

        ego_plan = out["output_trajectory"][0][..., :2]
        joint_plan = ego_plan[:, :self.branch_horizon].contiguous()
        sep_plan = ego_plan[:, None, self.branch_horizon:].repeat(1, self.modal, 1, 1).contiguous()

        batch = joint_plan.shape[0]

        planner_inputs = {
            "joint_plan": joint_plan.reshape(batch, -1),
            "sep_plan": sep_plan.reshape(batch, -1),
            "actors": actor_pred,
            "actors_hw": actor_shape,
            "mode_score": actor_score,
            "actor_mask": actor_mask,
            "original_plan":ego_plan
        }

        index = torch.tensor([[i for i in range(self.modal)]], dtype=torch.float32).to(self.device)
        for i in range(self.modal):
            planner_inputs[f'index_{i}'] = index[:, None, i]

        if not self.test:
            final_values, _ = self.layer.forward(planner_inputs)
            joint_plan, sep_plan = final_values["joint_plan"], final_values["sep_plan"]
        else:
            _, info = self.layer.forward(planner_inputs, optimizer_kwargs={'track_best_solution': True,  "verbose": True})
            joint_plan = info.best_solution['joint_plan'].to(self.device)
            sep_plan = info.best_solution['sep_plan'].to(self.device)
        
        joint_plan = joint_plan.view(-1, self.branch_horizon, 2)
        joint_plan = joint_plan[:, None].repeat(1, self.modal, 1, 1)
        sep_plan = sep_plan.view(-1, self.modal, self.horizon - self.branch_horizon, 2)

        contigency_plan = torch.cat([joint_plan, sep_plan], dim=-2)
        return contigency_plan
    
    def plan_simple(self, data):
        
        actor_shape = data["shape"][:, :, :2]
        actor_curr_pos = data["position"][:, :, :2]

        actor_preds = data['prediction']
        actor_score = data['pred_probability']

        print(
            actor_shape.shape,
            actor_curr_pos.shape,
            actor_preds.shape,
            actor_score.shape
        )

        actor_pred_pos = actor_preds[..., :2] + actor_curr_pos[:, :, None, None]
        actor_angle = actor_preds[..., 2]
        actor_pred = torch.cat([actor_pred_pos, actor_angle.unsqueeze(-1)], dim=-1)

        actor_mask = torch.ones((1, 1)).to(actor_pred.device)
        actor_mask = actor_mask.float()

        ego_plan = data["output_trajectory"][..., :2]
        joint_plan = ego_plan[:, :self.branch_horizon].contiguous()
        sep_plan = ego_plan[:, None, self.branch_horizon:].repeat(1, self.modal, 1, 1).contiguous()

        batch = joint_plan.shape[0]

        planner_inputs = {
            "joint_plan": joint_plan.reshape(batch, -1),
            "sep_plan": sep_plan.reshape(batch, -1),
            "actors": actor_pred,
            "actors_hw": actor_shape,
            "mode_score": actor_score,
            "actor_mask": actor_mask,
            "original_plan":ego_plan
        }

        index = torch.tensor([[i for i in range(self.modal)]], dtype=torch.float32).to(self.device)
        for i in range(self.modal):
            planner_inputs[f'index_{i}'] = index[:, None, i]

        if not self.test:
            final_values, _ = self.layer.forward(planner_inputs)
            joint_plan, sep_plan = final_values["joint_plan"], final_values["sep_plan"]
        else:
            _, info = self.layer.forward(planner_inputs, optimizer_kwargs={'track_best_solution': True, "verbose": True})
            joint_plan = info.best_solution['joint_plan'].to(self.device)
            sep_plan = info.best_solution['sep_plan'].to(self.device)
        
        joint_plan = joint_plan.view(-1, self.branch_horizon, 2)
        joint_plan = joint_plan[:, None].repeat(1, self.modal, 1, 1)
        sep_plan = sep_plan.view(-1, self.modal, self.horizon - self.branch_horizon, 2)

        contigency_plan = torch.cat([joint_plan, sep_plan], dim=-2)
        return contigency_plan
```
